crates/sui-core/src/histogram.rs

```diff
@@ -0,0 +1,389 @@
+    start: Instant,
+}
+
+#[derive(Clone)]
+pub struct HistogramVec {
+    channel: mpsc::UnboundedSender<HistogramMessage>,
+}
+
+struct HistogramCollector {
+    reporter: Arc<Mutex<HistogramReporter>>,
+    channel: mpsc::UnboundedReceiver<HistogramMessage>,
+}
+
+struct HistogramReporter {
+    gauge: IntGaugeVec,
+    sum: IntCounterVec,
+    count: IntCounterVec,
+    known_labels: HashSet<HistogramLabels>,
+    percentiles: Vec<usize>,
+}
+
+type HistogramLabels = Arc<HistogramLabelsInner>;
+
+struct HistogramLabelsInner {
+    labels: Vec<String>,
+    hash: u64,
+}
+
+/// Reports the histogram to the given prometheus gauge.
+/// Unlike the histogram from prometheus crate, this histogram does not require to specify buckets
+/// It works by calculating 'true' histogram by aggregating and sorting values.
+///
+/// The values are reported into prometheus gauge with requested labels and additional dimension
+/// for the histogram percentile.
+///
+/// It worth pointing out that due to those more precise calculations, this Histogram usage
+/// is somewhat more limited comparing to original prometheus Histogram.
+///
+/// On the bright side, this histogram exports less data to Prometheus comparing to prometheus::Histogram,
+/// it exports each requested percentile into separate prometheus gauge, while original implementation creates
+/// gauge per bucket.
+/// It also exports _sum and _count aggregates same as original implementation.
+///
+/// It is ok to measure timings for things like network latencies and expensive crypto operations.
+/// However as a rule of thumb this histogram should not be used in places that can produce very high data point count.
+///
+/// As a last round of defence this histogram emits error log when too much data is flowing in and drops data points.
+///
+/// This implementation puts great deal of effort to make sure the metric does not cause any harm to the code itself:
+/// * Reporting data point is a non-blocking send to a channel
+/// * Data point collections tries to clear the channel as fast as possible
+/// * Expensive histogram calculations are done in a separate blocking tokio thread pool to avoid effects on main scheduler
+/// * If histogram data is produced too fast, the data is dropped and error! log is emitted
+impl HistogramVec {
+    pub fn new_in_registry(name: &str, desc: &str, labels: &[&str], registry: &Registry) -> Self {
+        Self::new_in_registry_with_percentiles(
+            name,
+            desc,
+            labels,
+            registry,
+            vec![500usize, 950, 990],
+        )
+    }
+
+    /// Allows to specify percentiles in 1/1000th, e.g. 90pct is specified as 900
+    pub fn new_in_registry_with_percentiles(
+        name: &str,
+        desc: &str,
+        labels: &[&str],
+        registry: &Registry,
+        percentiles: Vec<usize>,
+    ) -> Self {
+        let sum_name = format!("{}_sum", name);
+        let count_name = format!("{}_count", name);
```

Comment:
Perhaps we can add a counter "_drop" which tracks if the histogram is dropping samples

Reply:
I thought about it, however it is a bit hard since it would need to copy counter to many places where drop can happen(those two counters are only stored in the reporter). Overall I think reporting with `error!` log should be good enough, I think there should be some counter counter number of errors